guides/micronaut-java-records/micronaut-java-records.adoc

```diff
@@ -0,0 +1,193 @@
+
+Configure the database migrations directory for http://www.liquibase.org[Liquibase] in `application.yml`.
+
+resource:application.yml[tag=liquibase]
+
+Create the following files with the database schema creation:
+
+resource:db/liquibase-changelog.xml[] 
+
+resource:db/changelog/01-create-books-schema.xml[] 
+
+During application startup, http://www.liquibase.org[Liquibase] executes the SQL file and create the schema needed for the application.
+
+=== Mapped Entities with Java Records
+
+Create a Micronaut Data Mapped Entity
+
+source:Book[]
+
+callout:mapped-entity[1]
+<2> The primary key is annotated with `@Id`. It is an assigned primary key.
+callout:constraints[3]
+<4> Annotate `about` with `@Nullable` since it is optional.
+
+=== Projections with Java Records
+
+Create a record to project some data from the `book` table. For example, exclude the `about` field.
+
+source:BookCard[]
+
+callout:introspected[1]
+callout:constraints[2]
+
+Create a https://micronaut-projects.github.io/micronaut-data/latest/guide/#dbcRepositories[Repository] which uses the previous Java record as a DTO projection.
+
+source:BookRepository[]
+
+callout:jdbcrepository[1]
+callout:crudrepository[2]
+callout:dto-projection[3]
+
+=== JSON serialization with Java Records
+
+Create a Java record to represent a JSON response: 
+
+source:BookForSale[]
+
+callout:introspected[1]
+callout:constraints[2]
+
+Create a Controller which uses the previous record: 
+
+source:BookController[]
+
+callout:controller[number=1,arg0=/books]
+callout:constructor-di[number=2,arg0=BookRepository]
+callout:get[number=3,arg0=index,arg1='/']
+<4> Respond a Java Record and the application responds JSON representation of the object
+
+=== Testing with PostgreSQL with TestContainers
+
+We use https://www.testcontainers.org/modules/databases/postgres/[Test Containers] to test against a PostgreSQL database. 
+
+Add the following snippet to include the necessary test container dependencies:
+
+:dependencies:
+dependency:postgresql[groupId=org.testcontainers,scope=testImplementation]
+dependency:testcontainers[groupId=org.testcontainers,scope=testImplementation]
+:dependencies:
+
+Configure the default datasource to use the PostgreSQL database provided by TestContainers for the test environment:
+
+testResource:application-test.yml[]
+
+=== Create a Test
+
+Create a test:
+
+test:BookControllerTest[]
+
+callout:property[1]
+callout:micronaut-test-transactional-false[2]
+callout:http-client[3]
+<4> Multiline string 
+callout:binding-json-array[5]
+
+=== Running the application
+
+include::{commondir}/common-run-postgres-with-docker.adoc[]
+
+Set up the following environment variables to connect to the PostgreSQL database you started with Docker.
+
+resource:application.yml[tag=datasource]
+
+<1> The JDBC url matches the database name you used in the previous command.
+<2> Driver and dialect are configured to use PostgreSQL.
+<3> You handle database migrations via Liquibase 
```

Comment:
Another callout is not rendered

![image](https://user-images.githubusercontent.com/559192/142607729-46238d38-d753-4e0b-817b-addac39c73ce.png)